src/details/ArborX_DetailsDistributedTreeNearest.hpp

```diff
@@ -186,119 +191,142 @@ struct CallbackWithDistance<BoundingVolumeHierarchy<
+                                 Predicates const &predicates,
+                                 Distances &farthest_distances)
 {
-  Kokkos::Profiling::ScopedRegion guard(
-      "ArborX::DistributedTree::query::nearest::deviseStrategy");
+  std::string prefix = "ArborX::DistributedTree::query::nearest::phaseI";
+  Kokkos::Profiling::ScopedRegion guard(prefix);
 
   using namespace DistributedTree;
   using MemorySpace = typename Tree::memory_space;
 
-  auto const &top_tree = tree._top_tree;
-  auto const &bottom_tree_sizes = tree._bottom_tree_sizes;
+  auto comm = tree.getComm();
 
-  // Find the k nearest local trees.
-  top_tree.query(space, predicates, LegacyDefaultCallback{}, nearest_ranks,
-                 offset);
+  // Find the k nearest local trees
+  Kokkos::View<int *, MemorySpace> offset(prefix + "::offset", 0);
+  Kokkos::View<int *, MemorySpace> nearest_ranks(prefix + "::nearest_ranks", 0);
+  tree._top_tree.query(space, predicates, nearest_ranks, offset);
 
-  // Accumulate total leave count in the local trees until it reaches k which
-  // is the number of neighbors queried for.  Stop if local trees get
-  // empty because it means that they are no more leaves and there is no point
-  // on forwarding predicates to leafless trees.
   auto const n_predicates = predicates.size();
-  Kokkos::View<int *, MemorySpace> new_offset(
-      Kokkos::view_alloc(space, offset.label()), n_predicates + 1);
-  Kokkos::parallel_for(
-      "ArborX::DistributedTree::query::nearest::"
-      "bottom_trees_with_required_cumulated_leaves_count",
-      Kokkos::RangePolicy<ExecutionSpace>(space, 0, n_predicates),
-      KOKKOS_LAMBDA(int i) {
-        int leaves_count = 0;
-        int const n_nearest_neighbors = getK(predicates(i));
-        for (int j = offset(i); j < offset(i + 1); ++j)
-        {
-          int const bottom_tree_size = bottom_tree_sizes(nearest_ranks(j));
-          if ((bottom_tree_size == 0) || (leaves_count >= n_nearest_neighbors))
-            break;
-          leaves_count += bottom_tree_size;
-          ++new_offset(i);
-        }
-      });
 
-  KokkosExt::exclusive_scan(space, new_offset, new_offset, 0);
+  // Accumulate total leave count in the local trees until it reaches k which
+  // is the number of neighbors queried for
+  {
+    auto const &bottom_tree_sizes = tree._bottom_tree_sizes;
+    auto new_offset =
+        KokkosExt::cloneWithoutInitializingNorCopying(space, offset);
+    auto new_nearest_ranks =
+        KokkosExt::cloneWithoutInitializingNorCopying(space, nearest_ranks);
+    int new_num_nearest_ranks;
+    Kokkos::parallel_scan(
+        "ArborX::DistributedTree::query::nearest::"
+        "bottom_trees_with_required_cumulated_leaves_count",
+        Kokkos::RangePolicy<ExecutionSpace>(space, 0, n_predicates),
+        KOKKOS_LAMBDA(int i, int &update, bool final_pass) {
+          auto const k = getK(predicates(i));
+          if (k > 0)
```

Comment:
Interestingly, we have a distributed test where this is not true (`k = comm_rank * comm_size = 0`).